flow/adaptive_throttler.go

```diff
@@ -0,0 +1,413 @@
+		CPUUsageMode:        CPUUsageModeMeasured,   // Use actual process CPU usage via gopsutil
+		HysteresisBuffer:    5.0,                    // Prevent oscillations around threshold
+		MaxRateChangeFactor: 0.3,                    // More conservative rate changes
+	}
+}
+
+// resourceMonitorInterface defines the interface for resource monitoring
+type resourceMonitorInterface interface {
+	GetStats() ResourceStats
+	IsResourceConstrained() bool
+	Close()
+}
+
+// AdaptiveThrottler implements a feedback control system that:
+// - Monitors CPU and memory usage at regular intervals
+//
+// - Reduces throughput when resources exceed thresholds (with severity-based scaling)
+//
+// - Gradually increases throughput when resources are available (with hysteresis)
+//
+// - Applies smoothing to prevent abrupt rate changes
+//
+// - Enforces minimum and maximum throughput bounds
+type AdaptiveThrottler struct {
+	config  AdaptiveThrottlerConfig
+	monitor resourceMonitorInterface
+
+	// Current rate (elements per second)
+	currentRate atomic.Int64
+
+	// Rate control
+	period      time.Duration // Calculated from currentRate
+	maxElements atomic.Int64  // Elements per period
+	counter     atomic.Int64
+
+	// Channels
+	in          chan any
+	out         chan any
+	quotaSignal chan struct{}
+	done        chan struct{}
+
+	// Rate adaptation
+	adaptMu        sync.Mutex
+	lastAdaptation time.Time
+
+	stopOnce sync.Once
+}
+
+// Validate checks that the configuration is valid and returns an error if not
+func (c *AdaptiveThrottlerConfig) Validate() error {
+	if c.MaxMemoryPercent <= 0 || c.MaxMemoryPercent > 100 {
+		return fmt.Errorf("invalid MaxMemoryPercent: %f", c.MaxMemoryPercent)
+	}
+	if c.MinThroughput < 1 || c.MaxThroughput < c.MinThroughput {
+		return fmt.Errorf("invalid throughput bounds: min=%d, max=%d", c.MinThroughput, c.MaxThroughput)
+	}
+	if c.AdaptationFactor <= 0 || c.AdaptationFactor >= 1 {
+		return fmt.Errorf("AdaptationFactor must be in (0, 1), got %f", c.AdaptationFactor)
+	}
+	if c.MaxCPUPercent <= 0 || c.MaxCPUPercent > 100 {
+		return fmt.Errorf("invalid MaxCPUPercent: %f", c.MaxCPUPercent)
+	}
+	if c.BufferSize < 1 {
+		return fmt.Errorf("invalid BufferSize: %d", c.BufferSize)
+	}
+	if c.SampleInterval <= 0 {
+		return fmt.Errorf("invalid SampleInterval: %v", c.SampleInterval)
+	}
+	if c.HysteresisBuffer < 0 {
+		return fmt.Errorf("invalid HysteresisBuffer: %f", c.HysteresisBuffer)
+	}
+	if c.MaxRateChangeFactor <= 0 || c.MaxRateChangeFactor > 1 {
+		return fmt.Errorf("invalid MaxRateChangeFactor: %f, must be in (0, 1]", c.MaxRateChangeFactor)
+	}
+	return nil
+}
+
+// Verify AdaptiveThrottler satisfies the Flow interface
+var _ streams.Flow = (*AdaptiveThrottler)(nil)
+
+// NewAdaptiveThrottler creates a new adaptive throttler
+func NewAdaptiveThrottler(config AdaptiveThrottlerConfig) *AdaptiveThrottler {
+	if err := config.Validate(); err != nil {
+		panic(fmt.Sprintf("invalid config: %v", err))
```

Comment:
Document all the panics.

Reply:
Not returning errors enables the seamless inlining of new flows into the pipeline.
Given the wide range of possible errors, returning the error would be better in this particular case.